File: template/node/start_server.js
// start_server.js
// to run: node start_server.js

// start_server: v4

const http = require('http');
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');
const { escape } = require('html-escaper');
const querystring = require('querystring');
const { DateTime } = require('luxon');

class CustomHTTPRequestHandler {
	constructor(req, res) {
		this.req = req;
		this.res = res;
		this.directory = process.cwd();
	}

	handleRequest() {
		if (this.req.method === 'GET') {
			this.handleGet();
		} else if (this.req.method === 'POST') {
			this.handlePost();
		} else {
			this.sendError(405, "Method Not Allowed");
		}
	}

	handleGet() {
		if (this.req.url === '/') {
			this.serveFile('index.html');
		} else if (this.req.url === '/log.html') {
			this.checkAndGenerateReport();
			this.serveFile('log.html');
		} else if (this.req.url === '/chat.html') {
			this.checkAndGenerateChatHtml();
			this.serveFile('chat.html');
		} else if (this.req.url === '/api/github_update') {
			this.handleGithubUpdate();
		} else if (this.req.url.endsWith('.txt')) {
			this.serveTextFile();
		} else {
			this.serveFile(this.req.url.slice(1));
		}
	}

	handlePost() {
		if (this.req.url === '/chat.html') {
			this.handleChatPost();
		} else {
			this.sendError(405, "Method Not Allowed");
		}
	}

	handleChatPost() {
		let body = '';
		this.req.on('data', chunk => {
			body += chunk.toString();
		});
		this.req.on('end', () => {
			const params = querystring.parse(body);
			const author = params.author || '';
			const message = params.message || '';

			if (author && message) {
				this.saveMessage(author, message);
				this.res.writeHead(200, { 'Content-Type': 'text/html' });
				this.res.end("Message saved successfully" + '<meta http-equiv="refresh" content="1;url=chat.html">');

				// Commit the message and update GitHub
				this.runScript('commit_files', 'message');
				this.runScript('github_update');
			} else {
				this.sendError(400, "Bad Request: Missing author or message");
			}
		});
	}

	saveMessage(author, message) {
		const today = DateTime.now().toFormat('yyyy-MM-dd');
		const directory = path.join(this.directory, 'message', today);
		fs.mkdirSync(directory, { recursive: true });

		const title = this.generateTitle(message);
		const filename = `${title}.txt`;
		const filepath = path.join(directory, filename);

		fs.writeFileSync(filepath, `${message}\n\nauthor: ${author}`, 'utf-8');
	}

	generateTitle(message) {
		let title = message.split(' ').slice(0, 5).join('_');
		title = title.replace(/[^a-zA-Z0-9_-]/g, '');
		if (!title) {
			title = Math.random().toString(36).substring(2, 12);
		}
		return title;
	}

	handleGithubUpdate() {
		this.res.writeHead(200, { 'Content-Type': 'text/html' });
		this.res.end("Update triggered successfully");
		this.runScript('github_update');
	}

	checkAndGenerateReport() {
		const htmlFile = path.join(this.directory, 'log.html');
		if (!fs.existsSync(htmlFile) || Date.now() - fs.statSync(htmlFile).mtime.getTime() > 60000) {
			console.log(`${htmlFile} is older than 60 seconds or does not exist. Running log.html script...`);
			this.runScript('log.html');
		} else {
			console.log(`${htmlFile} is up-to-date.`);
		}
	}

	checkAndGenerateChatHtml() {
		const chatHtmlFile = path.join(this.directory, 'chat.html');
		if (!fs.existsSync(chatHtmlFile) || Date.now() - fs.statSync(chatHtmlFile).mtime.getTime() > 60000) {
			console.log(`${chatHtmlFile} is older than 60 seconds or does not exist. Running chat.html script...`);
			this.runScript('chat.html');
		} else {
			console.log(`${chatHtmlFile} is up-to-date.`);
		}
	}

	runScript(scriptName, ...args) {
		const scriptTypes = ['py', 'js', 'rb', 'sh', 'pl'];
		const interpreters = ['python3', 'node', 'ruby', 'bash', 'perl'];
		const interpreterMap = Object.fromEntries(scriptTypes.map((type, index) => [type, interpreters[index]]));

		const foundScripts = [];

		// Search for scripts in template/*/script_name.*
		fs.readdirSync(path.join(this.directory, 'template')).forEach(dir => {
			scriptTypes.forEach(type => {
				const fullPath = path.join(this.directory, 'template', dir, `${scriptName}.${type}`);
				if (fs.existsSync(fullPath)) {
					foundScripts.push(fullPath);
				}
			});
		});

		if (foundScripts.length === 0) {
			console.log(`No scripts found for ${scriptName}`);
			return;
		}

		foundScripts.forEach(script => {
			const type = path.extname(script).slice(1);
			if (interpreterMap[type]) {
				const interpreter = interpreterMap[type];
				console.log(`Running ${script} with ${interpreter}...`);

				// Escape arguments to prevent shell injection
				const escapedArgs = args.map(arg => `"${arg.replace(/"/g, '\\"')}"`);
				const argString = escapedArgs.join(' ');

				exec(`${interpreter} ${script} ${argString}`, (error, stdout, stderr) => {
					if (error) console.error(`Error: ${error.message}`);
					if (stderr) console.error(`Error: ${stderr}`);
					console.log(`Output: ${stdout}`);
				});
			} else {
				console.log(`No suitable interpreter found for ${script}`);
			}
		});
	}

	serveTextFile() {
		const filePath = path.join(this.directory, this.req.url.slice(1));
		fs.readFile(filePath, 'utf8', (err, content) => {
			if (err) {
				this.sendError(404, 'File not found');
				return;
			}

			const htmlContent = `
	<!DOCTYPE html>
	<html lang="en">
	<head>
	    <meta charset="UTF-8">
	    <meta name="viewport" content="width=device-width, initial-scale=1.0">
	    <title>${path.basename(filePath)}</title>
	    <style>
	        body { font-family: Arial, sans-serif; line-height: 1.6; padding: 20px; }
	        pre { background-color: #f4f4f4; padding: 15px; border-radius: 5px; white-space: pre-wrap; word-wrap: break-word; }
	    </style>
	</head>
	<body>
	    <h1>${path.basename(filePath)}</h1>
	    <pre>${escape(content)}</pre>
	</body>
	</html>
            `;

			this.res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
			this.res.end(htmlContent);
		});
	}

	serveFile(fileName) {
		const filePath = path.join(this.directory, fileName);
		fs.readFile(filePath, (err, content) => {
			if (err) {
				this.sendError(404, 'File not found');
				return;
			}

			const ext = path.extname(filePath).toLowerCase();
			const contentType = {
				'.html': 'text/html',
				'.js': 'text/javascript',
				'.css': 'text/css',
				'.json': 'application/json',
				'.png': 'image/png',
				'.jpg': 'image/jpeg',
				'.gif': 'image/gif',
			}[ext] || 'application/octet-stream';

			this.res.writeHead(200, { 'Content-Type': contentType });
			this.res.end(content);
		});
	}

	sendError(statusCode, message) {
		this.res.writeHead(statusCode, { 'Content-Type': 'text/plain' });
		this.res.end(message);
	}
}

function isPortInUse(port) {
	return new Promise((resolve) => {
		const server = http.createServer();
		server.once('error', () => {
			resolve(true);
		});
		server.once('listening', () => {
			server.close();
			resolve(false);
		});
		server.listen(port);
	});
}

async function findAvailablePort(startPort) {
	let port = startPort;
	while (await isPortInUse(port)) {
		port++;
	}
	return port;
}

async function runServer(port, directory) {
	process.chdir(directory);
	const server = http.createServer((req, res) => {
		const handler = new CustomHTTPRequestHandler(req, res);
		handler.handleRequest();
	});

	try {
		await new Promise((resolve, reject) => {
			server.listen(port, () => {
				console.log(`Serving HTTP on 0.0.0.0 port ${port} (http://0.0.0.0:${port}/) ...`);
				resolve();
			});
			server.on('error', reject);
		});
		return true;
	} catch (error) {
		if (error.code === 'EADDRINUSE') {
			console.log(`Port ${port} is already in use.`);
			return false;
		}
		throw error;
	}
}

function parseArguments() {
	const args = process.argv.slice(2);
	let port;
	let directory = process.cwd();

	for (let i = 0; i < args.length; i += 2) {
		if (args[i] === '-p' || args[i] === '--port') {
			port = parseInt(args[i + 1]);
		} else if (args[i] === '-d' || args[i] === '--directory') {
			directory = args[i + 1];
		}
	}

	return { port, directory };
}

async function main() {
	const { port: specifiedPort, directory } = parseArguments();

	if (specifiedPort) {
		if (!await runServer(specifiedPort, directory)) {
			console.log(`Failed to start server on specified port ${specifiedPort}.`);
		}
	} else {
		let port = 8000;
		while (!await runServer(port, directory)) {
			port = await findAvailablePort(port + 1);
			console.log(`Trying port ${port}...`);
		}
	}
}

main().catch(console.error);

// end of start_server.js

File: template/ruby/start_server.rb
#!/usr/bin/env ruby

# start_server.rb
# to run: ruby start_server.rb

# start_server: v4

require 'webrick'
require 'optparse'
require 'fileutils'
require 'uri'
require 'cgi'
require 'date'
require 'socket'

class CustomHTTPServer < WEBrick::HTTPServlet::AbstractServlet
  def initialize(server, directory)
    super(server)
    @directory = directory
  end

  def do_GET(request, response)
    case request.path
    when '/'
      serve_file(response, 'index.html')
    when '/log.html'
      check_and_generate_report
      serve_file(response, 'log.html')
    when '/chat.html'
      check_and_generate_chat_html
      serve_file(response, 'chat.html')
    when '/api/github_update'
      handle_github_update(response)
    when /\.txt$/
      serve_text_file(response, request.path)
    else
      serve_file(response, request.path[1..-1])
    end
  end

  def do_POST(request, response)
    if request.path == '/chat.html'
      handle_chat_post(request, response)
    else
      response.status = 405
      response.body = "Method Not Allowed"
    end
  end

  private

  def check_and_generate_report
    html_file = File.join(@directory, 'log.html')
    if !File.exist?(html_file) || Time.now - File.mtime(html_file) > 60
      puts "#{html_file} is older than 60 seconds or does not exist. Running log.html.rb..."
      run_script('log.html')
    else
      puts "#{html_file} is up-to-date."
    end
  end

  def check_and_generate_chat_html
    chat_html_file = File.join(@directory, 'chat.html')
    if !File.exist?(chat_html_file) || Time.now - File.mtime(chat_html_file) > 60
      puts "#{chat_html_file} is older than 60 seconds or does not exist. Running chat.html.rb..."
      run_script('chat.html')
    else
      puts "#{chat_html_file} is up-to-date."
    end
  end

  def handle_github_update(response)
    response.status = 200
    response.body = "Update triggered successfully"
    run_script('github_update')
  end

  def handle_chat_post(request, response)
    params = CGI.parse(request.body)
    author = CGI.unescape(params['author'].first || '')
    message = CGI.unescape(params['message'].first || '')

    if author.empty? || message.empty?
      response.status = 400
      response.body = "Bad Request: Missing author or message"
    else
      save_message(author, message)
      response.status = 200
      response.content_type = 'text/html'
      response.body = "Message saved successfully" + '<meta http-equiv="refresh" content="1;url=chat.html">'
      run_script('commit_files', 'message')
      run_script('github_update')
    end
  end

  def save_message(author, message)
    today = Date.today.strftime("%Y-%m-%d")
    dir = File.join(@directory, 'message', today)
    FileUtils.mkdir_p(dir)

    title = generate_title(message)
    filename = "#{title}.txt"
    filepath = File.join(dir, filename)

    File.open(filepath, 'w') do |file|
      file.puts "#{message}\n\nauthor: #{author}"
    end
  end

  def generate_title(message)
    title = message.split.first(5).join('_')
    title.gsub!(/[^a-zA-Z0-9_-]/, '')
    title = SecureRandom.hex(5) if title.empty?
    title
  end

  def serve_file(response, filename)
    file_path = File.join(@directory, filename)
    if File.exist?(file_path)
      response.body = File.read(file_path)
      response.content_type = get_content_type(file_path)
      response.status = 200
    else
      response.status = 404
      response.body = "File not found"
    end
  end

  def serve_text_file(response, path)
    file_path = File.join(@directory, path[1..-1])
    if File.exist?(file_path)
      content = File.read(file_path)
      html_content = generate_html_for_text_file(File.basename(file_path), content)
      response.body = html_content
      response.content_type = 'text/html; charset=utf-8'
      response.status = 200
    else
      response.status = 404
      response.body = "File not found"
    end
  end

  def generate_html_for_text_file(filename, content)
    content = CGI.escapeHTML(content)
    <<-HTML
    <!DOCTYPE html>
    <html lang="en">
    <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <title>#{filename}</title>
      <style>
        body { font-family: Arial, sans-serif; line-height: 1.6; padding: 20px; }
        pre { background-color: #f4f4f4; padding: 15px; border-radius: 5px; white-space: pre-wrap; word-wrap: break-word; }
      </style>
    </head>
    <body>
      <h1>#{filename}</h1>
      <pre>#{content}</pre>
    </body>
    </html>
    HTML
  end

  def get_content_type(file)
    case File.extname(file)
    when '.txt' then 'text/plain'
    when '.html' then 'text/html'
    when '.css' then 'text/css'
    when '.js' then 'application/javascript'
    when '.json' then 'application/json'
    when '.png' then 'image/png'
    when '.jpg', '.jpeg' then 'image/jpeg'
    when '.gif' then 'image/gif'
    else 'application/octet-stream'
    end
  end

  def run_script(script_name, *args)
    script_types = ['py', 'rb', 'pl', 'sh', 'js']
    interpreters = ['python3', 'ruby', 'perl', 'bash', 'node']
    interpreter_map = Hash[script_types.zip(interpreters)]

    found_scripts = Dir.glob("template/*/#{script_name}.*")

    if found_scripts.empty?
      puts "No scripts found for #{script_name}"
      return
    end

    found_scripts.each do |script|
      type = File.extname(script)[1..-1]
      if interpreter_map.key?(type)
        interpreter = interpreter_map[type]
        puts "Running #{script} with #{interpreter}..."
        escaped_args = args.map { |arg| Shellwords.escape(arg) }
        system("#{interpreter} #{script} #{escaped_args.join(' ')}")
      else
        puts "No suitable interpreter found for #{script}"
      end
    end
  end
end

def is_port_in_use?(port)
  begin
    server = TCPServer.new('127.0.0.1', port)
    server.close
    false
  rescue Errno::EADDRINUSE
    true
  end
end

def find_available_port(start_port)
  port = start_port
  port += 1 while is_port_in_use?(port)
  port
end

options = {
  port: 8000,
  directory: Dir.pwd
}

OptionParser.new do |opts|
  opts.banner = "Usage: ruby start_server.rb [options]"

  opts.on("-p", "--port PORT", Integer, "Port to run the server on") do |p|
    options[:port] = p
  end

  opts.on("-d", "--directory DIR", "Directory to serve files from") do |d|
    options[:directory] = d
  end
end.parse!

Dir.chdir(options[:directory])

if is_port_in_use?(options[:port])
  puts "Port #{options[:port]} is already in use."
  options[:port] = find_available_port(options[:port] + 1)
  puts "Trying port #{options[:port]}..."
end

server = WEBrick::HTTPServer.new(
  Port: options[:port],
  DocumentRoot: options[:directory],
  BindAddress: '0.0.0.0'
)

server.mount "/", CustomHTTPServer, options[:directory]

trap('INT') { server.shutdown }

puts "Serving HTTP on 0.0.0.0 port #{options[:port]} (http://0.0.0.0:#{options[:port]}/) ..."
server.start